lib/systems/examples.nix

```diff
@@ -93,6 +93,26 @@ rec {
     config = "mipsel-unknown-linux-gnu";
   } // platforms.fuloong2f_n32;
 
+  # MIPS ABI table transcribed from here: https://wiki.debian.org/Multiarch/Tuples
+
+  # can execute on 32bit chip
+  mips-linux-gnu                = { config = "mips-linux-gnu";                } // platforms.gcc_mips32r2_o32;
```

Comment:
This this been tested at all? This does not even build stdenv: https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/76

Reply:
These are not yet supported, nor are they expected to evaluate.  From [the commit message](https://github.com/NixOS/nixpkgs/pull/161158/commits/12371a51e647a00b90fe250837f056642125c095),

```
MIPS has a large space of {architecture,abi,endianness}; this commit
adds all of them to lib/systems/platforms.nix so we can be done with
it.
```

As the one-line comment says, it is just a transcription of the table which enumerates all the possible {cpu,kernel,libc,abi} combinations, since mips has a lot of them -- probably more than any other architecture.  I transcribed the table to (a) discourage future platform support from cooking up custom nonstandard platform names and (b) show people that no, there isn't an infinite neverending list of combinations; there is a limit to how many of these we will ever care about.

>  https://hercules-ci.com/github/nix-community/cross-toolchains.nix/jobs/76

> It seems that you've disabled JavaScript on this page. The Hercules CI dashboard needs that though.

:stuck_out_tongue_closed_eyes:

Reply:
I don't see the point of having a tuple that cannot be tested in any way. This sounds like deadcode to me. It should have at least a comment stating that it cannot be used yet.

Reply:
There are approximately 182 licenses in `licenses.nix`, none of which can be "tested in an way".  This is okay, because they aren't code; they are us teaching Nix about a list of things that exists outside of Nix, in the real world.

The mips 5-tuple table is the same thing, but for chips instead of legal documents.  Don't think of it as executable code.

Sorry to see that you [got frustrated so quickly](https://github.com/NixOS/nixpkgs/pull/165020#issuecomment-1073697553) with trying to add a new platform to nixpkgs.  I found the experience very frustrating too.

I have also found the nixpkgs process for merging complex, multi-commit changes (like the mips64el support) to be frustrating.  Reviewers very sensibly ask that large and complex changes be broken up into multiple PRs, but GitHub totally fails to provide any kind of dependency tracking between bugs.  Since PRs that do not evaluate correctly tend to not attract any reviews, the person submitting a complex multi-PR changeset basically has to wait for each one to be merged before submitting the next.  Technically you can submit them all at once, like I did, but when you have a long dependency chain only the first PR in the chain will evaluate correctly; the rest will not evaluate (due to the earlier one not yet being merged), so those will fail CI and not be reviewed until the early one gets merged and then the submitter goes back and manually rebases all the later PRs.  So really there is no benefit to submitting them all at once.  The submitter has to manually submit them one at a time, waiting for each to merge before submitting the next.

To summarize: I think the nixpkgs-github process is optimized to attract "drive-by" one-liner patches from casual contributors (like my systemdSupport patches), at the cost of not being able to cope well with large/complex changes (like my mips64el patches) and refactorings (like cleaning up lib/systems/, which is... tangled).  I'm sure that optimizing for casual contributions helped nixpkgs grow fast early on, but it doesn't seem to be a good way to limit the growth of "technical debt".

Anyways, because of the headaches of managing not-yet-fully-merged multi-PR submissions, I'm going to wait for all of my mips64el work to make it all the way to master, and bootstrap tarballs to build, before I start any work on the rest of the mips tuples.  The little-endian ones will be first; in order to test big-endian I'll need to set up another machine with a big-endian kernel.